File: dotnet/Aula-2018-09-25/CounterSemaphoreFIFO.cs
using SynchUtils;
using System;
using System.Collections.Generic;
using System.Threading;

namespace Aula_2018_09_25 {
    public class CounterSemaphoreFIFO {
        private Object monitor;
        private int permits;
        LinkedList<int> waiters;

        public CounterSemaphoreFIFO(int initialPermits) {
            if (initialPermits >= 0)
                permits = initialPermits;
            monitor = new Object();
            waiters = new LinkedList<int>();
        }

        private void notifyWaiters() {
            if (waiters.Count > 0 && permits >= waiters.First.Value)  
                Monitor.PulseAll(monitor);
        }

        public bool Acquire(int requests, int timeout) { // throws InterruptedException  
            lock (monitor) {
                // non blocking path
                if (permits >= requests && waiters.Count > 0) {
                    permits -= requests;
                    return true;
                }
                if (timeout == 0)
                    return false;
                // blocking path
                // prepare wait
                TimeoutHolder th = new TimeoutHolder(timeout);
                var node = waiters.AddLast(requests);
                try {
                    do {
                        Monitor.Wait(monitor, th.Value);
                        if (waiters.First == node &&  permits >= requests) {
                            waiters.RemoveFirst();
                            permits -= requests;
                            notifyWaiters();
                            return true;
                        }
                        if (th.Timeout) {
                            waiters.Remove(node);
                            notifyWaiters();
                            return false;
                        }
                    }
                    while (true);
                }
                catch (ThreadInterruptedException e) {
                    waiters.Remove(node);
                    notifyWaiters();
                    throw e;
                }
            }
        }

        public void Release(int releases) {
            lock (monitor) {
                permits += releases;
                Monitor.PulseAll(monitor);
            }
        }
    }
}

File: java/src/isel/leic/pc/monitors/CounterSemaphoreFIFO.java
package isel.leic.pc.monitors;

import isel.leic.pc.utils.TimeoutHolder;

import java.util.LinkedList;


public class CounterSemaphoreFIFO {
    private Object monitor;
    private int permits;

    private static class Request {
        private final int n;
        private Request(int n) { this.n = n;}
    }
    private LinkedList<Request> waiters;

    public CounterSemaphoreFIFO(int initialPermits) {
        if (initialPermits >=0)
            permits = initialPermits;
        monitor = new Object();
        waiters = new LinkedList<>();
    }

    private void notifyWaiters() {
        if (waiters.size() > 0 && permits >= waiters.getFirst().n)
            monitor.notifyAll();
    }

    public boolean acquire(int n, long timeout) throws InterruptedException {
        synchronized(monitor) {
            // Here is the correct test to do.
            // On the lecture, the test for empty waiters list was forgotten,
            // which permits barging, that is, is possible that an
            // acquiring thread passes in front of waiting threads,
            // jeopardizing FIFO discipline!
            if (permits >= n && waiters.size() == 0) {
                permits -= n;
                return true;
            }
            if (timeout == 0) return false;
            TimeoutHolder th = new TimeoutHolder(timeout);
            Request req = new Request(n);
            waiters.add(req);

            try {
                do {
                    monitor.wait(th.value());
                    if (waiters.getFirst() == req && permits >= n) {
                        waiters.removeFirst();
                        permits -= n;
                        notifyWaiters();
                        return true;
                    }
                    if (th.value() == 0) {
                        waiters.remove(req);
                        notifyWaiters();
                        return false;
                    }
                }
                while (true);
            }
            catch(InterruptedException e) {
                waiters.remove(req);
                notifyWaiters();
                throw e;
            }
        }
    }

    public void release(int n) {
        synchronized (monitor) {
            permits += n;
            notifyWaiters();
        }
    }

}
